[선을로/익조] 로또 3단계 - 수동구매 기능 추가

ikjo93 opened the pull request:
안녕하세요!! hounx!! 15팀(선을로/익조)입니다!!

"로또 2단계 - 보너스 번호 추가" 구현 부분과 관련하여, 작성한 소스 코드파일을 첨부드리오니 확인부탁드립니다. :)

※ 참고사항 : 클래스 역할과 책임, 이벤트 흐름도 등은 README 파일을 참고해주세요!!

감사합니다.

ikjo93 committed:
feat: 사용자 수동으로 추첨 번호 입력 및 통계 산출 기능 구현

수정사항
1. UserInterface 클래스를 를 통해 사용자로부터 수동으로 구매할 로또 개수 및 번호를 입력 받도록 하고, 이를 통해 자동 로또 개수와 수동 로또 개수를 구했습니다.
2. 당초 LottoMaker 클래스에서는 자동으로 로또번호를 생성하는 기능만 있었지만, 수동으로 로또 번호를 생성하는 기능을 추가하도록 하였습니다. 이에 따라 변수명과 메서드명의 변화를 주었습니다.(makeAutoLotto() 메서드, makeManualLotto() 메서드 등)
3. 이와 관련하여 InputView 클래스와 OutputView 클래스의 출력문 등을 수정하였습니다.

기타
1. 당초 lottoCount 와 같은 변수명이 명확하지 않음(콩글리쉬?)으로 countOfLotto 와 같은 올바른 영문 표기법의 변수명을 지향하였습니다.
2. RankValue enum에서 valueOf 메서드에서 당첨 번호 개수가 3개 미만인 경우(4개 if문에 해당되지 않는 경우) NOTHING을 반환하도록 했습니다.

ikjo93 committed:
feat : 사용자 입력 검증 기능 구현

수정사항
1. UserInterface 객체에서 Scanner를 통해 사용자 입력을 받은 결과를 Validation 객체에 전달해주어 이를 검증(예외처리)하도록 했고 올바르지 않은 입력인 경우 다시 입력받도록 구현했습니다.

기타
1. 당초 LottoGame에서 사용자 입력부를 별도 inputLottoInfo 메서드에 분리시켜놓았었는데 논리적으로 적합하지 않아 다시 run 메서드에 포함시켰습니다.
2. 당초 Lotto 클래스와 WinngingStatistic 클래스에서는 멤버로 가지고 있는 자료구조(List, Map)을 외부로 공개하고 있었는데 디미터의 법칙에 따라 해당 자료구조가 변경되지 않도록 해당 클래스에서 값을 처리하여 반환하도록 수정하였습니다. 이에 따라 OutputView 클래스 역시 수정했습니다.

ikjo93 committed:
docs : README 파일 생성

해당 프로그램(미션 3단계 기준) 클래스의 역할과 책임, 이벤트 흐름도 등에 대한 설명을 기입한 README 파일을 생성했습니다.

ikjo93 committed:
refactor : 기존 LottoMaker를 AutoLottoMaker와 ManualLottoMaker로 분리

수정사항
1. 인터페이스 LottoMaker를 생성하고 이를 구현하는 AutoLottoMaker 클래스와 ManualLottoMaker 클래스를 생성하였습니다.
2. 이에 따라 LottoGame 클래스 등의 코드를 수정하였습니다.

ikjo93 committed:
refactor : private 접근 제어자 누락분 추가

tmdgusya commented on AutoLottoMaker.java:
```diff
@@ -0,0 +1,37 @@
+        return list;
+    }
+
+    private List<Integer> makeLottoNumber() {
+        Collections.shuffle(totalNumberList);
+        List<Integer> lottoNumberList = new ArrayList<>();
+
+        for (int i = 0; i < 6; i++) {
+            lottoNumberList.add(totalNumberList.get(i));
+        }
+        Collections.sort(lottoNumberList);
+
+        return lottoNumberList;
+    }
```
저번 리뷰에서 말한 부분을 적용하셨군요
근데 한가지 궁금한 점이 있습니다. 
혹시 makeLottoNumber 자체가 interface 의 메소드여야 하지 않을까요?
둘다 makeLottoNumber 가 있을거 같아서요 만약 다른 클래스에 노출하기 싫다면 protected 를 쓸 수 있는 abstract class 도 사용을 고려해보시는 것도.

어렵겠지만 만약 추상클래스 였다면 어땠을지 한번 생각해보세요.

ikjo93 replied:
당초 LottoMaker 인터페이스를 통해 AutoLottoMaker와 ManualLottoMaker를 구현하고있었으나, 리뷰주신 사항과 같이 makeLottoNumber 메서드 역시도 LottoMaker를 구현하는 메서드로서 Override됨이 마땅하며, 아울러 이들간의 관계(LottoMaker ↔ Auto/ManualLottoMaker)는 is a kind of(~의 종류인) 관계로서 인터페이스 보다는 추상 클래스트를 통해 구현하는 것이 더 적합하다고 판단하여 추상클래스 형태로 변경 작업하였습니다.

tmdgusya commented on ManualLottoMaker.java:
```diff
@@ -0,0 +1,33 @@
+        return list;
+    }
+
+    private List<Integer> makeLottoNumber(String manualLottoNumber) {
```
이 interface 가 겹친다는 생각이 드시진 않나요 ㅎㅎ?
추상 클래스 한번 고려해보시죠

ikjo93 replied:
위에서 답변드린 사항과 동일합니다. 😁

<br/>

```
당초 LottoMaker 인터페이스를 통해 AutoLottoMaker와 ManualLottoMaker를 구현하고있었으나, 리뷰주신 사항과 같이 makeLottoNumber 메서드 역시도 LottoMaker를 구현하는 메서드로서 Override됨이 마땅하며, 아울러 이들간의 관계(LottoMaker ↔ Auto/ManualLottoMaker)는 is a kind of(~의 종류인) 관계로서 인터페이스 보다는 추상 클래스트를 통해 구현하는 것이 더 적합하다고 판단하여 추상클래스 형태로 변경 작업하였습니다.
```

tmdgusya commented on RankValue.java:
```diff
@@ -28,6 +29,6 @@ public static RankValue valueOf(int countOfMatch, boolean matchBonus) {
         if (countOfMatch == FIRST.countOfMatch) return FIRST;
 
-        return null;
+        return NOTHING;
```
👍

ikjo93 replied:
🙇‍♂️

tmdgusya commented on UserInterface.java:
```diff
@@ -4,27 +4,98 @@
+        while (true) {
+            InputView.guideInputMoney();
+            money = validation.validateMoney(scanner.nextLine());
+            if (money > -1) break;
```
-1 은 무슨 뜻일까요 ㅎㅎ?
validateMoney 에서 int 가 나오나요?

ikjo93 replied:
당초 Validation 클래스에서 사용자 입력을 검증 시 어떤 메서드는 int로(당초 의도는 문자열 형태인 사용자 데이터를 정수로 반환하고자함이었습니다.) 그리고 어떤 메서드는 String으로 반환하는 등 반환 타입이 중구난방이었으나, 검증 메서드의 특성상 반환형은 boolean형이 더 적합하다고 판단하여 Validation 메서드의 반환형을 모두 boolean형으로 수정하였습니다.

tmdgusya commented on Validation.java:
```diff
@@ -0,0 +1,72 @@
+
+    private static final int LOTTO_EACH_MONEY = 1000;
+
+    public int validateMoney(String userInput) {
```
보통 `is` 나 `validate` 에서는 boolean 값을 기대하지 않을까요?

tmdgusya replied:
이 부분은 한번 고려를 해보셔야 할거 같아요

ikjo93 replied:
위에서 답변드린 사항과 동일합니다. 😁

<br/>

```
당초 Validation 클래스에서 사용자 입력을 검증 시 어떤 메서드는 int로(당초 의도는 문자열 형태인 사용자 데이터를 정수로 반환하고자함이었습니다.) 그리고 어떤 메서드는 String으로 반환하는 등 반환 타입이 중구난방이었으나, 검증 메서드의 특성상 반환형은 boolean형이 더 적합하다고 판단하여 Validation 메서드의 반환형을 모두 boolean형으로 수정하였습니다.
```

ikjo93 replied:
늦은 시간에도 리뷰 주셔서 감사드립니다!! ✨ 막상 Pull Request를 하고 나니 생각이 짧았던 부분들과 미흡했던 부분들이 많았습니다. 다음에 제출할 때는 좀 더 신경 써서 제출드리겠습니다. 😂 아울러 Changes requested를 안주셨으면 이러한 부분들을 놓치고 그냥 지나쳤을 것 같았는데 다시 한 번 생각하게 해주셔서 감사드립니다. 🙏

tmdgusya reviewed (changes requested):
메소드 명에 기대되는 결과값에 아쉬운 부분이 있어 
해당 부분은 메소드 명을 수정하거나 바꿔주셔야 할거 같아요.

추상클래스 관련된 부분은 한번 고려해보시는걸 추천드려요 ㅎㅎ
지금 상태로 가도 좋으나 좀 더 좋은 방법이 있어보입니다.
그래도 좋은 고민을 하신 노력들이 보여요 👍

ikjo93 committed:
refactor : 코드 리뷰 피드백 반영 및 로직 개선 작업

수정사항
1. 당초 LottoMaker 인터페이스를 통해 AutoLottoMaker와 ManualLottoMaker를 구현하고있었으나, 이들간의 관계는 is a kind of(~의 종류인)이 더 적합하여 인터페이스 보다는 추상 클래스트를 통해 구현하는 것이 더 적합하다고 판단하여 추상클래스 형태로 변경 작업하였습니다.
2. 당초 Validation 클래스에서 사용자 입력을 검증 시 어떤 메서드는 int 어떤 메서드는 String으로 반환하는 등 반환 타입이 중구난방이었으나, 검증 메서드의 특성상 반환형은 boolean형이 더 적합하다고 판단하여 반환형을 모두 boolean형으로 수정하였고 이에 따라 UserInterface의 로직을 수정하였습니다.

기타
1. 당초 다수의(2개 이상) 수동 로또 번호를 입력받을 시 만일 사용자 입력 오류(예외) 발생 시 이전에 기록하고있었던 수동 로또 번호가 모두 초기화되어 다시 처음부터 입력해야 하는 등 좋지 않은 UX를 제공하고 있었는데, 이전 수동 로또 번호를 기억하도록 로직을 개선했습니다. ex) 수동 로또 번호 2개 → 첫번째 번호 "정상" 입력 → 두번째 번호 "비정상" 입력 → "다시" 두번째 번호 입력(첫번째 번호는 기억하고 있음)

ikjo93 commented:
늦은 시간에도 리뷰주셔서 감사드립니다!! ✨ 막상 Pull Request를 하고 나니 생각이 짧았던 부분들과 미흡했었던 부분들이 많았습니다. 다음부터는 제출하기 전에 좀 더 신경써서 제출드리겠습니다. 😂 아울러 만일 Changes requested가 아니었다면 생각하지 않고 그냥 넘어갔을 수도 있는 부분들이었는데 다시 한 번 생각할 수 있게 해주셔서 감사드립니다!! 🙏

tmdgusya commented:
> 

@ikjo93 
ㅎㅎ 페어로 잘 작성하시는거 같고 두분이서 학습을 잘 하시는 것 같네요.
아마 이번에 클래스를 나눠보시면서 많은 학습을 하셨을거 같네요.
수정하신거 올려주시면 머지해드리겠습니다.

ikjo93 commented:
@tmdgusya
감사합니다 로치님!! 이번 리뷰를 통해 저희 15팀 많이 배우게 되었습니다~! 🙏

ikjo93 commented:
@tmdgusya
로치님 😭 저희가 git 사용법이 미숙해서(😂) 실수로 Merge Pull Request를 눌러서 저희 이름으로 Merge가 된 거 같습니다 🤣 수정한 내용은 Origin에 Push 해놓은 상태였었는데(커밋 기록 9010a5b) 혹시 다시 검토를 받아 Merge를 받도록 해야한다면 Revert 요청 해놓겠습니다..!!

tmdgusya commented on LottoMaker.java:
```diff
@@ -1,6 +1,8 @@
 import java.util.List;
 
-public interface LottoMaker {
+public abstract class LottoMaker {
```
👍

tmdgusya commented on LottoMaker.java:
```diff
@@ -1,6 +1,8 @@
-    List<Lotto> getLottoList(int countOfLotto);
+    public abstract List<Lotto> getLottoList(int countOfLotto);
+
+    protected abstract List<Integer> makeLottoNumber();
```
제가 생각한것과 비슷하군요. 좋습니다 :)

시간 되시면 템플릿 메소드 패턴에 대해서 한번 훑어만 보세요. 따로 공부는 안하셔도 됩니다 ㅎㅎ
보시면 디자인 패턴에 대해서 감이 오실꺼에요.

tmdgusya commented on Validation.java:
```diff
@@ -7,58 +7,56 @@
 
-    public int validateBonusNumber(String userInput) {
-        int bonusNumber = -1;
+    public boolean validateBonusNumber(String userInput) {
```
👍

tmdgusya reviewed (commented):
리뷰 반영하느라 고생하셨습니다.
추상 클래스로 잘 바꾸신거 같아 기분이 좋네요 💯
